### PR TITLE
refactor: Use ESLint and FlatESLint from lib/api.js

### PR DESCRIPTION
**Pull request by bartlomieju**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[ ] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[x] Other, please explain:
Preparation for bundling "api.js" as part of the build process

#### What changes did you make? (Give an overview)

Replaced two `require` calls for `ESLint` and `FlatESLint` in `lib/cli.js` with a single call to obtain them from `lib/api.js`. Consequently "FlatESLint" is now being reexported from "lib/api.js".

#### Is there anything you'd like reviewers to focus on?

Prerequisite for https://github.com/eslint/eslint/issues/16820 and addresses the comment https://github.com/eslint/eslint/issues/16820#issuecomment-1440517534.

Let me know if you'd prefer me to do all the changes related to bundling in a single PR.

### CODE REVIEW

File: lib/cli.js [hunk 1]
```diff
@@ -18,8 +18,7 @@
 const fs = require("fs"),
     path = require("path"),
     { promisify } = require("util"),
-    { ESLint } = require("./eslint"),
-    { FlatESLint } = require("./eslint/flat-eslint"),
+    { ESLint, FlatESLint } = require("./eslint"),
```
Comment (nzakas) [inline lib/cli.js:21]: Based on the previous change I'd expect you really meant `require("./api")` here?

Comment (bartlomieju) [reply lib/cli.js:21]: Yes, of course 🤦

File: lib/api.js [hunk 2]
```diff
@@ -21,6 +22,7 @@ const { SourceCode } = require("./source-code");
 module.exports = {
     Linter,
     ESLint,
+    FlatESLint,
```
[the remark this answers is not in the document]
Comment (bartlomieju) [reply lib/api.js:25]: That's a good question - let me do some testing locally to see if I can see any difference in performance.

[the remark this answers is not in the document]
Comment (bartlomieju) [reply lib/api.js:25]: @nzakas here are some benchmarks.

The first one is using bundled `api.js` file, then second one is running straight from `main` branch. Both of them are using default ESLint config for JavaScript; linting a single, empty JavaScript file.

```
❯ ../deno/third_party/prebuilt/mac/hyperfine --warmup 10 -S none 'node ../eslint/bin/eslint.js ./empty.js'
Benchmark 1: node ../eslint/bin/eslint.js ./empty.js
  Time (mean ± σ):     180.7 ms ±   2.7 ms    [User: 182.5 ms, System: 18.6 ms]
  Range (min … max):   177.5 ms … 186.8 ms    16 runs


~/dev/bench_eslint via  v19.6.0 on ☁️  bartek@deno.com took 4s
❯ ../deno/third_party/prebuilt/mac/hyperfine --warmup 10 -S none 'node ../eslint/bin/eslint.js ./empty.js'
Benchmark 1: node ../eslint/bin/eslint.js ./empty.js
  Time (mean ± σ):     152.8 ms ±   0.8 ms    [User: 153.6 ms, System: 18.5 ms]
  Range (min … max):   151.6 ms … 154.6 ms    19 runs

```

We can see here that bundled version is actually taking 30ms more to execute. My gut feeling was that even though we're loading bundled `api.js` there is still some code that is loading dependencies of `api.js` - effectively executing the same code twice. Adding a `console.log` to at the top of `lib/linter/linter.js` confirms it. 

I think the first step would be to ensure the dependency graph between the various ESLint files is somewhat "simplified" - ie. there are some "barrel" modules that are meant to be public interface to various part of the codebase and that there are no side imports that circumvent these "barrel" modules. 

I might be completely wrong here, so please correct me if so.

[the remark this answers is not in the document]
Comment (nzakas) [reply lib/api.js:25]: I'm sure that's the case -- we never really paid much attention to which modules were including what, so no doubt that there are multiple paths getting to the same files.

I'm also not sure that it's worth the time to go digging through to figure this out. As we progress in the rewrite, we'll be creating a cleaner codebase where we can plan from the start to have a better structure that is more conducive to bundling. Especially once we are able to remove the eslintrc config completely, that will clean up the structure significantly.

What do you think?

[the remark this answers is not in the document]
Comment (bartlomieju) [reply lib/api.js:25]: That is a perfectly reasonable solution in my eyes. This works doesn't really block me on my side project and I can proceed with a custom bundle in the meantime. 

I'm going to close the PR for now. Let me know if I can help you folks in any other capacity!

**Pull request closed without merging**